controllers/rollback_handlers.go

```diff
@@ -18,18 +18,90 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"path/filepath"
 
-	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
 	"github.com/openshift-kni/lifecycle-agent/controllers/utils"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"github.com/openshift-kni/lifecycle-agent/internal/common"
+	"github.com/openshift-kni/lifecycle-agent/internal/ostreeclient"
+	lcautils "github.com/openshift-kni/lifecycle-agent/utils"
 	ctrl "sigs.k8s.io/controller-runtime"
+
+	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 //nolint:unparam
-func (r *ImageBasedUpgradeReconciler) handleRollback(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+func (r *ImageBasedUpgradeReconciler) startRollback(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	utils.SetRollbackStatusInProgress(ibu, "Initiating rollback")
+
+	stateroot, err := r.RPMOstreeClient.GetUnbootedStaterootName()
+	if err != nil {
+		utils.SetRollbackStatusFailed(ibu, err.Error())
+		return requeueWithError(err)
+	}
+
+	if _, err := r.Executor.Execute("mount", "/sysroot", "-o", "remount,rw"); err != nil {
+		return requeueWithError(err)
+	}
+
+	stateRootRepo := common.PathOutsideChroot(ostreeclient.StaterootPath(stateroot))
+
+	// Save the CR for post-reboot restore, stripping Upgrade conditions and ResourceVersion from IBU CR before saving
+	r.Log.Info("Save the IBU CR to the old state root before pivot")
+	filePath := filepath.Join(stateRootRepo, utils.IBUFilePath)
+	// Temporarily empty resource version so the file can be used to restore status
+	rv := ibu.ResourceVersion
+	ibu.ResourceVersion = ""
+	if err := lcautils.MarshalToFile(ibu, filePath); err != nil {
+		return requeueWithError(err)
+	}
+	ibu.ResourceVersion = rv
+
+	r.Log.Info("Finding unbooted deployment")
+	deploymentIndex, err := r.RPMOstreeClient.GetUnbootedDeploymentIndex()
+	if err != nil {
+		utils.SetRollbackStatusFailed(ibu, err.Error())
+		return requeueWithError(fmt.Errorf("failed to get unbooted deployment index: %w", err))
+	}
+
+	// Set the new default deployment
+	r.Log.Info("Checking for set-default feature")
+
+	if r.OstreeClient.IsOstreeAdminSetDefaultFeatureEnabled() {
```

Comment:
Logic seems very similar between both functions, can we somehow unite it?

Reply:
Do you mean the if/else clauses? These are very different paths. If set-default is available, the rollback handler will use it to set the new default deployment automatically. If it is not, it checks to see if the user has manually set it, and fails the rollback otherwise, with a message reminding the user to manually set it. And at some point, the "set-default not available" code path will be removed, once we define a minimum-required release.